dnf5/commands/search/search_processor.cpp

```diff
@@ -0,0 +1,173 @@
+    : base(base),
+      patterns(patterns),
+      search_all(search_all),
+      full_package_query(base),
+      showdupes(base.get_config().showdupesfromrepos().get_value()) {
+    if (!showdupes) {
+        full_package_query.filter_latest_evr();
+    }
+}
+
+libdnf::rpm::PackageSet SearchProcessor::get_matches(
+    const std::string & pattern, int priority, filter_method_type filter) {
+    auto cmp = get_cmp_from_pattern(pattern);
+
+    libdnf::rpm::PackageQuery query_contains(full_package_query);
+    (query_contains.*filter)({pattern}, cmp);
+    update_priorities(query_contains, priority);
+
+    if (patterns.size() == 1 && cmp != libdnf::sack::QueryCmp::IGLOB) {
+        libdnf::rpm::PackageQuery query_exact(query_contains);
+        (query_exact.*filter)({pattern}, libdnf::sack::QueryCmp::EXACT);
+        update_priorities(query_exact, (priority << 1));
+    }
+
+    return query_contains;
+}
+
+libdnf::rpm::PackageSet SearchProcessor::get_name_matches(const std::string & pattern) {
+    return get_matches(pattern, get_search_key_priority(SEARCH_KEY_NAME), &libdnf::rpm::PackageQuery::filter_name);
+}
+
+libdnf::rpm::PackageSet SearchProcessor::get_summary_matches(const std::string & pattern) {
+    return get_matches(
+        pattern, get_search_key_priority(SEARCH_KEY_SUMMARY), &libdnf::rpm::PackageQuery::filter_summary);
+}
+
+libdnf::rpm::PackageSet SearchProcessor::get_description_matches(const std::string & pattern) {
+    return get_matches(
+        pattern, get_search_key_priority(SEARCH_KEY_DESCRIPTION), &libdnf::rpm::PackageQuery::filter_description);
+}
+
+libdnf::rpm::PackageSet SearchProcessor::get_url_matches(const std::string & pattern) {
+    return get_matches(pattern, get_search_key_priority(SEARCH_KEY_URL), &libdnf::rpm::PackageQuery::filter_url);
+}
+
+void SearchProcessor::update_priorities(const libdnf::rpm::PackageSet & packages, int priority) {
+    for (auto const & package : packages) {
+        packages_priorities[package.get_full_nevra()] |= priority;
+    }
+}
+
+SearchResults SearchProcessor::get_results() {
+    libdnf::rpm::PackageSet all_matches(base);
+
+    for (auto const & pattern : patterns) {
+        libdnf::rpm::PackageSet pattern_matches(base);
+        pattern_matches |= get_name_matches(pattern);
+        pattern_matches |= get_summary_matches(pattern);
+
+        if (search_all) {
+            pattern_matches |= get_description_matches(pattern);
+            pattern_matches |= get_url_matches(pattern);
+        }
+
+        // For the first pattern we are always adding to the empty list.
+        // If we are using the "--all" option, we want any matches in the result.
+        // Otherwise we use an intersection (AND) with all previous results.
+        if (all_matches.empty() | search_all) {
+            all_matches.update(std::move(pattern_matches));
+        } else {
+            all_matches.intersection(std::move(pattern_matches));
+        }
```

Comment:
When the first pattern isn't matched, `all_matches` remains empty, but intersection should still be used for any subsequent pattern. E.g. `dnf5 search foobarbaz dnf` should result in an empty output, but doesn't (provided `foobarbaz` is something non-existent).

Similarly, when an intersection results in empty `all_matches`, intersection is not used for the subsequent pattern. E.g. `dnf5 search glibc nano dnf` should result in an empty output, but doesn't.

Reply:
Oh, I see. I should distinguish going through the first loop instead of checking the empty matches. Thanks!